http-middleware/security.go

```diff
@@ -0,0 +1,224 @@
+					logFields.WithError(err).Warn("User is not authorized")
+					http_model.WriteJSONResponse(w, http.StatusUnauthorized, http_model.ErrResponseUnauthorized)
+					return
+				}
+			}
+
+			span.End()
+			next.ServeHTTP(w, req)
+		})
+	}
+}
+
+func handleAccessToken(req *http.Request, header string) error {
+	base64Token := req.Header.Get(header)
+	if base64Token == "" {
+		return errors.Errorf("auth header [%s] was empty", header)
+	}
+
+	token, err := jwt.Parse(base64Token)
+	if err != nil {
+		return errors.New("authorization token not valid")
+	}
+
+	*req = *req.WithContext(
+		context.WithValue(req.Context(), claimsContextKey{}, token.GetClaims()),
+	)
+	return nil
+}
+
+// ExtractClaimsFromContext extracts JWT claims from a context.
+func ExtractClaimsFromContext(ctx context.Context) (_ jwt.Claims, err error) {
+	v := ctx.Value(claimsContextKey{})
+	if v == nil {
+		err = errors.New("unable to parse Claims from context")
+		return
+	}
+
+	claims := v.(jwt.Claims)
+	return claims, nil
+}
+
+type Authorizer interface {
+	IsAuthorizedWithContext(ctx context.Context, userID, action string, resource *common.Origin) (bool, error)
+}
+
+// AuthorizeMiddleware retrieves the security configuration for the matched route
+// and handles the configured authorizations.
+func AuthorizeMiddleware(authorizer Authorizer) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx, span := trace.StartSpan(req.Context(), "Authorizer")
+			defer span.End()
+			*req = *req.WithContext(ctx)
+
+			logFields := log.
+				WithField("method", req.Method).
+				WithField("url", req.URL.String())
+
+			secConfig := lookupSecurityConfig(req)
+			if len(secConfig.authorizations) == 0 {
+				span.End()
+				next.ServeHTTP(w, req)
+				return
+			}
+
+			claims, err := ExtractClaimsFromContext(req.Context())
+			if err != nil {
+				logFields.Error("Couldn't extract claims from context.")
+				http_model.WriteJSONResponse(w, http.StatusInternalServerError, http_model.ErrResponseInternalServerError)
+				return
+			}
+
+			for _, authorizeConfig := range secConfig.authorizations {
+				resource, err := authorizeConfig.resourceFunc(req)
+				if err != nil {
+					logFields.WithError(err).Error("ResourceFunc failed.")
+					http_model.WriteJSONResponse(w, http.StatusInternalServerError, http_model.ErrResponseInternalServerError)
+					return
+				}
+
+				ok, err := authorizer.IsAuthorizedWithContext(
+					ctx,
+					claims.UserID,
+					authorizeConfig.action,
+					resource,
+				)
+				if !ok || err != nil {
+					logFields.
+						WithField("userId", claims.UserID).
+						WithField("action", authorizeConfig.action).
+						WithField("resource", resource).
+						Warn("User is not Authorized")
+					http_model.WriteJSONResponse(w, http.StatusUnauthorized, http_model.ErrResponseUnauthorized)
+					return
```

Comment:
Check for missing span.End()s

Reply:
I added a `defer` for the ones where we plainly `return` and don't run `next.ServeHTTP(w, req)`